File: payment/src/main/java/dev/orderlab/payment/ServiceInfoController.java
package dev.orderlab.payment;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/internal")
class ServiceInfoController {

    private final String serviceName;

    ServiceInfoController(@Value("${spring.application.name}") String serviceName) {
        this.serviceName = serviceName;
    }

    @GetMapping("/service-info")
    ServiceInfoResponse serviceInfo() {
        var thread = Thread.currentThread();
        return new ServiceInfoResponse(
                serviceName,
                "UP",
                Runtime.version().toString(),
                thread.getName(),
                thread.isVirtual());
    }

    record ServiceInfoResponse(
            String service,
            String status,
            String javaVersion,
            String threadName,
            boolean virtualThread) {
    }
}

File: pricing/src/main/kotlin/dev/orderlab/pricing/ServiceInfoController.kt
package dev.orderlab.pricing

import org.springframework.beans.factory.annotation.Value
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/internal")
class ServiceInfoController(
    @Value("\${spring.application.name}") private val serviceName: String,
) {

    @GetMapping("/service-info")
    fun serviceInfo(): ServiceInfoResponse {
        val thread = Thread.currentThread()
        return ServiceInfoResponse(
            service = serviceName,
            status = "UP",
            javaVersion = Runtime.version().toString(),
            threadName = thread.name,
            virtualThread = thread.isVirtual,
        )
    }
}

data class ServiceInfoResponse(
    val service: String,
    val status: String,
    val javaVersion: String,
    val threadName: String,
    val virtualThread: Boolean,
)
